### ISSUE TITLE
[ffi.compile] Sizes are not properly resolved when a referenced type X has a pointer to type which includes X

### ISSUE THREAD

**Issue by droptopx** [python-cffi/cffi#147]
The following files cause a crash due to

`TypeError: field 'struct Outer.inner' has ctype 'struct Inner' of unknown size`

when they are run in sequence:

`gen_cffi_test.py`:
```python
import cffi

ffi = cffi.FFI()
ffi.set_source("bug_repro", None)
ffi.cdef(
    """\
struct Inner {
  struct Outer *outer_ref;
};

struct Outer {
  struct Inner inner;
};
"""
)
ffi.compile(verbose=True, debug=True, tmpdir="./bug_repro_autogen")
```

`cffi_test.py`:
```python
from bug_repro_autogen.bug_repro import ffi

ffi.callback("void (struct Inner *)")
```

`python gen_cffi_test.py && python cffi_test.py`

I have not figured out what the root cause is and will probably not be able to find time to look further. Any analysis by someone else is welcome.

EDIT: It's not just `ffi.callback`, `ffi.sizeof("struct Inner")` also gives the same error.
EDIT2: The error doesn't happen when the erroring method is called in the first file (i.e. no compilation)

**Comment by arigo** [comment]
It works for me and I remember fixing something like that (maybe last year?).  Can you check which version of CFFI you're testing?

**Comment by arigo** [comment]
Sorry, I got it to fail too now.  My mistake.  It only occurs in out-of-line ABI mode, when you specify `None` for the C sources in `ffi.set_source()`.

**Comment by arigo** [comment]
There is an order dependency, which suggests a workaround for you for now: if you first do `ffi.sizeof("struct Outer")` then any following usage of `struct Inner` works fine.

**Comment by arigo** [comment]
See https://github.com/python-cffi/cffi/pull/148.

**Comment by droptopx** [comment]
@arigo #148 does fix #147, thanks a lot for the fix and the great turn-around time!
